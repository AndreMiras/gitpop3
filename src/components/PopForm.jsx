import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
  Button, Form, FormControl, InputGroup,
} from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { urlMatch } from '../utils/validators';

const SearchIcon = ({ loading }) => (
  <FontAwesomeIcon icon={loading ? 'spinner' : 'search'} spin={loading} />
);
SearchIcon.propTypes = {
  loading: PropTypes.bool.isRequired,
};

const PopForm = ({ onSubmit, loading }) => {
  const [url, setUrl] = useState();
  const handleSubmit = (e) => {
    e.preventDefault();
    if (urlMatch(url) === null) {
      e.stopPropagation();
    } else {
      onSubmit(url);
    }
  };
  const isInvalid = url && urlMatch(url) === null;
  return (
    <Form onSubmit={handleSubmit}>
      <InputGroup className="mb-3">
        <FormControl
          placeholder="https://github.com/django/django"
          onChange={(e) => setUrl(e.target.value)}
          isInvalid={isInvalid}
        />
        <InputGroup.Append>
          <Button type="submit" variant="outline-secondary" onClick={handleSubmit}>
            <SearchIcon loading={loading} />
          </Button>
        </InputGroup.Append>
      </InputGroup>
    </Form>
  );
};
PopForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
};

export default PopForm;
