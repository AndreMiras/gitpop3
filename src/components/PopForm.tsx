import React, {
  FormEvent,
  FunctionComponent,
  useState,
  useEffect,
} from "react";
import { Button, Form, FormControl, InputGroup } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { urlMatch } from "../utils/validators";
import { IconProp } from "@fortawesome/fontawesome-svg-core";

const SearchIcon: FunctionComponent<{ loading: boolean }> = ({ loading }) => {
  const icon = loading ? ("spinner" as IconProp) : ("search" as IconProp);
  const spin = loading ? true : false;
  return <FontAwesomeIcon icon={icon} spin={spin} />;
};

type PopFormProps = {
  onSubmit: (url: string) => void;
  loading: boolean;
  initialUrl?: string;
};

const PopForm: FunctionComponent<PopFormProps> = ({
  onSubmit,
  loading,
  initialUrl,
}) => {
  const [url, setUrl] = useState(initialUrl || "");

  // Update local state if initialUrl prop changes
  useEffect(() => {
    if (initialUrl) {
      setUrl(initialUrl);
    }
  }, [initialUrl]);

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (urlMatch(url) === null) {
      e.stopPropagation();
    } else {
      onSubmit(url);
    }
  };

  const isInvalid = url !== "" && urlMatch(url) === null;

  return (
    <Form onSubmit={handleSubmit}>
      <InputGroup className="my-3">
        <FormControl
          placeholder="https://github.com/django/django"
          onChange={(e) => setUrl(e.target.value)}
          isInvalid={isInvalid}
          value={url}
        />
        <Button
          variant="primary"
          type="submit"
          disabled={loading}
          onClick={handleSubmit}
        >
          <SearchIcon loading={loading} />
        </Button>
      </InputGroup>
    </Form>
  );
};

export default PopForm;
