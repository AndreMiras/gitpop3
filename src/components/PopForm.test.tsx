import { describe, it, test, expect, vi } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render, screen } from "@testing-library/react";
import { library } from "@fortawesome/fontawesome-svg-core";
import { fas } from "@fortawesome/free-solid-svg-icons";
import PopForm from "./PopForm";

library.add(fas);

test("renders", () => {
  const onSubmit = () => ({});
  const tree = renderer
    .create(<PopForm onSubmit={onSubmit} loading={false} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe.each([true, false])("loading='%s'", (loading) => {
  test("renders", () => {
    const onSubmit = () => ({});
    const tree = renderer
      .create(<PopForm onSubmit={onSubmit} loading={loading} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

test("submit", () => {
  const onSubmit = vi.fn();
  render(<PopForm onSubmit={onSubmit} loading={false} />);
  const searchInput = screen.getByPlaceholderText(/github.com/);
  const expectedUrl = "https://github.com/django/django";
  fireEvent.change(searchInput, { target: { value: expectedUrl } });
  const submitButton = screen.getByRole("button");
  // mouse click
  fireEvent.click(submitButton);
  expect(onSubmit).toHaveBeenNthCalledWith(1, expectedUrl);
  // invalid URL should not trigger the submit callback
  const invalidUrl = "https://github.com/invalid-url";
  fireEvent.change(searchInput, { target: { value: invalidUrl } });
  fireEvent.click(submitButton);
  expect(onSubmit).toHaveBeenNthCalledWith(1, expectedUrl);
  // submit (keyboard Enter) should also be mapped
  fireEvent.change(searchInput, { target: { value: expectedUrl } });
  fireEvent.submit(searchInput);
  expect(onSubmit).toHaveBeenNthCalledWith(2, expectedUrl);
});

/**
 * Empty form should not trigger the submit callback
 */
test("submit empty", () => {
  const onSubmit = vi.fn();
  render(<PopForm onSubmit={onSubmit} loading={false} />);
  const searchInput = screen.getByPlaceholderText(/github.com/);
  const url = "";
  fireEvent.change(searchInput, { target: { value: url } });
  fireEvent.submit(searchInput);
  expect(onSubmit).not.toHaveBeenCalled();
});

test("renders with initial URL", () => {
  const onSubmit = vi.fn();
  const initialUrl = "https://github.com/django/django";
  render(
    <PopForm onSubmit={onSubmit} loading={false} initialUrl={initialUrl} />,
  );

  const searchInput = screen.getByPlaceholderText(
    /github.com/,
  ) as HTMLInputElement;
  expect(searchInput.value).toBe(initialUrl);
});

test("updates when initialUrl prop changes", () => {
  const onSubmit = vi.fn();
  const { rerender } = render(
    <PopForm
      onSubmit={onSubmit}
      loading={false}
      initialUrl="https://github.com/django/django"
    />,
  );

  let searchInput = screen.getByPlaceholderText(
    /github.com/,
  ) as HTMLInputElement;
  expect(searchInput.value).toBe("https://github.com/django/django");

  rerender(
    <PopForm
      onSubmit={onSubmit}
      loading={false}
      initialUrl="https://github.com/python/cpython"
    />,
  );

  searchInput = screen.getByPlaceholderText(/github.com/) as HTMLInputElement;
  expect(searchInput.value).toBe("https://github.com/python/cpython");
});
