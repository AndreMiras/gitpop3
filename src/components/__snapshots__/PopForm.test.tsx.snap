// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loading='false' > renders 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="mb-3 input-group"
  >
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="https://github.com/django/django"
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-outline-secondary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={{}}
          />
        </svg>
      </button>
    </div>
  </div>
</form>
`;

exports[`loading='true' > renders 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="mb-3 input-group"
  >
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="https://github.com/django/django"
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-outline-secondary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-spinner fa-w-16 fa-spin "
          data-icon="spinner"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
            fill="currentColor"
            style={{}}
          />
        </svg>
      </button>
    </div>
  </div>
</form>
`;

exports[`renders 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="mb-3 input-group"
  >
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="https://github.com/django/django"
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-outline-secondary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={{}}
          />
        </svg>
      </button>
    </div>
  </div>
</form>
`;
